### PR TITLE
[DSW-352] Add Data Management Plans Migration

### PR DESCRIPTION
**Pull request by josefdolezal**
This PR adds possibility to migrate existing questionnaires to newer knowledge model when its base model is outdated. Do not merge, it's still work in progress.

### CODE REVIEW

**Comment by josefdolezal** [comment]
The PR is ready for review now -- it works with latest server and implements diffing of integration questions ✅.

File: src/elm/Questionnaires/Index/View.elm [hunk 1]
```diff
@@ -130,12 +157,21 @@ listingActions wrapMsg appState questionnaire =
...
+        migrating =
+            questionnaire.state == Migrating
+
+        outdated =
+            questionnaire.state == Outdated
     in
     []
-        |> listInsertIf fillQuestionnaire editable
+        |> listInsertIf fillQuestionnaire (editable && not migrating)
         |> listInsertIf viewQuestionnaire (not editable)
```
Comment (janslifka) [inline src/elm/Questionnaires/Index/View.elm]: Maybe this one should also be restricted by migration? It is basically the same thing as fillQuestionnaire, except the questionnaire is rendered with disabled inputs.

File: src/elm/Questionnaires/Migration/Update.elm [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+            findQuestionReply replies replyPath
+
+        questionsChildren : List ( Int, List Question )
+        questionsChildren =
+            if List.length (getQuestionItemQuestions question) == 0 then
+                []
+
+            else
+                let
+                    -- We always want to see at least one item (even though it's not answered)
```
Comment (janslifka) [inline src/elm/Questionnaires/Migration/Update.elm]: I believe this is no longer true. Now, when there are no items in the questionnaire we only display the Add button. Is it somehow different in migrations?

File: src/elm/Questionnaires/Routing.elm [hunk 3]
```diff
@@ -24,6 +25,7 @@ parses wrapRoute =
     , map (wrapRoute << Detail) (s moduleRoot </> s "detail" </> string)
     , map (wrapRoute << Edit) (s moduleRoot </> s "edit" </> string)
     , map (wrapRoute <| Index) (s moduleRoot)
+    , map (wrapRoute << Migrate) (s moduleRoot </> string </> s "migration")
```
Comment (janslifka) [inline src/elm/Questionnaires/Routing.elm]: Why is this flipped, compared to everything else?

File: src/elm/Questionnaires/Routing.elm [hunk 4]
```diff
@@ -11,6 +11,7 @@ type Route
     | Detail String
     | Edit String
     | Index
+    | Migrate String
```
Comment (janslifka) [inline src/elm/Questionnaires/Routing.elm]: I would prefer to name it `Migration` to keep it consistent with the module name.

File: src/elm/Questionnaires/Migration/Update.elm (hunk 2 above)
Comment (josefdolezal) [reply src/elm/Questionnaires/Migration/Update.elm]: This is part of the questionnaire structure overview (the left panel). If the user answered by creating at leas one item, there will be a meta-note (marked as "Item 1", "Item 2" etc.) and if there is no answer, the questions from template will be shown directly under the template (see the attachment).
diff tree

**Comment by josefdolezal** [comment]
Alrighty, fixed the issues 👍

**Pull request closed without merging**